vpr/src/noc/turn_model_routing.h

```diff
@@ -0,0 +1,226 @@
+
+    /**
+     * @brief Returns the first horizontal direction found among given directions.
+     *
+     * @param directions  A list of directions.
+     *
+     * @return Direction The first horizontal direction found or INVALID if there
+     * is no horizontal direction among given directions.
+     */
+    TurnModelRouting::Direction select_horizontal_direction(const std::vector<TurnModelRouting::Direction>& directions);
+
+    /**
+     * @brief Returns the first direction among given direction
+     * that differs with "other_than" direction
+     *
+     * @param directions  A list of directions.
+     * @param other_than Specifies the direction that should not be returned.
+     *
+     * @return Direction The first direction that is different than "other_than"
+     * or INVALID if only "other_than" was among possible choices.
+     */
+    TurnModelRouting::Direction select_direction_other_than(const std::vector<TurnModelRouting::Direction>& directions,
+                                                            TurnModelRouting::Direction other_than);
+
+  private:
+    /**
+     * @brief Given the direction to travel next, this function determines
+     * the outgoing link that should be used to travel in the intended direction.
+     * Each router may have any number of outgoing links and each link is not
+     * guaranteed to point in the intended direction, So this function makes
+     * sure that the link chosen points in the intended direction.
+     *
+     * @param curr_router_id The physical router on the FPGA that the routing
+     * algorithm is currently visiting.
+     * @param curr_router_position The grid position of the router that is
+     * currently being visited on the FPGA
+     * @param next_step_direction The direction to travel next
+     * @param visited_routers Keeps track of which routers have been reached
+     * already while traversing the NoC.
+     * @param noc_model A model of the NoC. This is used to traverse the
+     * NoC and find a route between the two routers.
+     * @return NocLinkId A unique ID specifying the NoC link that should be
+     * traversed to move towards the given direction. If none of the current
+     * NoC router's links travel in the given direction, NoCLinkID::INVALID
+     * is returned.
+     */
+    NocLinkId move_to_next_router(NocRouterId& curr_router_id,
+                                  const t_physical_tile_loc& curr_router_position,
+                                  TurnModelRouting::Direction next_step_direction,
+                                  std::unordered_set<NocRouterId>& visited_routers,
+                                  const NocStorage& noc_model);
+
+
+    /**
+     * @brief Computes MurmurHash3 for an array of 32-bit words initialized
+     * with seed.
+     * @param key Contains elements to be hashed
+     * @param seed The initialization value.
+     * @return uint32_t Computed hash value.
+     */
```

Comment:
Briefly explain why this is used/helpful in turn routing. We use it with xxx to do yyy.